shell/apple/emulator-ios/emulator/EmulatorViewController.mm

```diff
@@ -156,135 +248,334 @@ - (void)update
+				if (dpad.right.isPressed) {
+					[self.emuView handleKeyDown:self.controllerView.img_dpad_r];
 				} else {
 					[self.emuView handleKeyUp:self.controllerView.img_dpad_r];
 				}
-				if (xValue <= -0.1) {
+				if (dpad.left.isPressed) {
 					[self.emuView handleKeyDown:self.controllerView.img_dpad_l];
 				} else {
 					[self.emuView handleKeyUp:self.controllerView.img_dpad_l];
 				}
-				if (yValue >= 0.1) {
+				if (dpad.up.isPressed) {
 					[self.emuView handleKeyDown:self.controllerView.img_dpad_u];
 				} else {
 					[self.emuView handleKeyUp:self.controllerView.img_dpad_u];
 				}
-				if (yValue <= -0.1) {
+				if (dpad.down.isPressed) {
 					[self.emuView handleKeyDown:self.controllerView.img_dpad_d];
 				} else {
 					[self.emuView handleKeyUp:self.controllerView.img_dpad_d];
 				}
             }];
-            [self.gController.extendedGamepad.leftThumbstick.xAxis setValueChangedHandler:^(GCControllerAxisInput *axis, float value){
-                
-            }];
-            [self.gController.extendedGamepad.leftThumbstick.yAxis setValueChangedHandler:^(GCControllerAxisInput *axis, float value){
-                
-            }];
+
+			[self.gController.gamepad.rightShoulder setValueChangedHandler:^(GCControllerButtonInput *button, float value, BOOL pressed) {
+				if (pressed) {
+					[self.emuView handleKeyDown:self.controllerView.img_rt];
+				} else {
+					[self.emuView handleKeyUp:self.controllerView.img_rt];
+				}
+			}];
+
+			[self.gController.gamepad.leftShoulder setValueChangedHandler:^(GCControllerButtonInput *button, float value, BOOL pressed) {
+				if (pressed) {
+					[self.emuView handleKeyDown:self.controllerView.img_lt];
+				} else {
+					[self.emuView handleKeyUp:self.controllerView.img_lt];
+				}
+			}];
+
+            //Add controller pause handler here
         }
     } else {
         self.gController = nil;
-//		[self.controllerView showController:self.view];
+		[self.controllerView showController:self.view];
     }
 }
 
+
 - (void)glkView:(GLKView *)view drawInRect:(CGRect)rect
 {
-    screen_width = view.drawableWidth;
+//    kdebug_signpost_start(10, 0, 0, 0, 0);
+	screen_width = view.drawableWidth;
     screen_height = view.drawableHeight;
 
     glClearColor(0.65f, 0.65f, 0.65f, 1.0f);
     glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
-    
-    while(!rend_single_frame()) ;
+
+	while(!rend_single_frame()) ;
+
+#if USE_CRT_SHADER
+	glUseProgram(crtShaderProgram);
+	glUniform4f(crtUniform_DisplayRect, 0,0, 320, 640); // TODO: Correct sizes?
+	glUniform1i(crtUniform_EmulatedImage, 0);
+	glUniform2f(crtUniform_EmulatedImageSize, 320, 640);
+	float finalResWidth = view.drawableWidth;
+	float finalResHeight = view.drawableHeight;
+	glUniform2f(crtUniform_FinalRes, finalResWidth, finalResHeight);
+#endif
+//	kdebug_signpost_end(10, 0, 0, 0, 0);
 }
 
+/*** CRT Shader borrowed from BrainDX for Provenance app ***/
+- (void)setupCRTShader
```

Comment:
I think in the future we want to have a more generic support for shader effects in the core, and not in the shell code. Looks okay though.